src/com/amazon/ion/impl/IonReaderContinuableApplicationBinary.java

```diff
@@ -0,0 +1,689 @@
+        // will need to be used.
+        return SharedSymbolTable.getSystemSymbolTable(getIonMajorVersion());
+    }
+
+    /**
+     * Reset the local symbol table to the system symbol table.
+     */
+    private void resetSymbolTable() {
+        // The following line is not required for correctness, but it frees the references to the old symbols,
+        // potentially allowing them to be garbage collected.
+        Arrays.fill(symbols, 0, localSymbolMaxId + 1, null);
+        localSymbolMaxId = -1;
+    }
+
+    /**
+     * Reset the list of imported shared symbol tables.
+     */
+    private void resetImports() {
+        // Note: when support for the next version of Ion is added, conditionals on 'majorVersion' and 'minorVersion'
+        // must be added here.
+        imports = ION_1_0_IMPORTS;
+        firstLocalSymbolId = imports.getMaxId() + 1;
+    }
+
+    /**
+     * Creates a shared symbol table import, resolving it from the catalog if possible.
+     * @param name the name of the shared symbol table.
+     * @param version the version of the shared symbol table.
+     * @param maxId the max_id of the shared symbol table. This value takes precedence over the actual max_id for the
+     *              shared symbol table at the requested version.
+     */
+    private SymbolTable createImport(String name, int version, int maxId) {
+        SymbolTable shared = catalog.getTable(name, version);
+        if (shared == null) {
+            // No match. All symbol IDs that fall within this shared symbol table's range will have unknown text.
+            return new SubstituteSymbolTable(name, version, maxId);
+        } else if (shared.getMaxId() != maxId || shared.getVersion() != version) {
+            // Partial match. If the requested max_id exceeds the actual max_id of the resolved shared symbol table,
+            // symbol IDs that exceed the max_id of the resolved shared symbol table will have unknown text.
+            return new SubstituteSymbolTable(shared, version, maxId);
+        } else {
+            // Exact match; the resolved shared symbol table may be used as-is.
+            return shared;
+        }
+    }
+
+    /**
+     * Retrieves the String text for the given symbol ID.
+     * @param sid a symbol ID.
+     * @return a String.
+     */
+    String getSymbol(int sid) {
+        if (sid < firstLocalSymbolId) {
+            return imports.findKnownSymbol(sid);
+        }
+        int localSid = sid - firstLocalSymbolId;
+        if (localSid > localSymbolMaxId) {
+            throw new IonException("Symbol ID exceeds the max ID of the symbol table.");
+        }
+        return symbols[localSid];
+    }
+
+    /**
+     * Uses the underlying raw reader to read the symbol tables from the stream. Capable of resuming if not enough
+     * data is currently available to complete the symbol table.
+     */
+    private class SymbolTableReader {
```

Comment:
If you wanted to, couldn't you require that the entire LST be in the buffer before you began reading? Then you wouldn't need the state machine.

Reply:
Yes, you could. The downside of that is that often the size of your symbol table is not closely related to the size of the user values in the stream (consider logging, where individual log events are small but have high cardinality). This makes it hard for users to choose a reasonable buffer configuration, especially because users tend to have less visibility into the size of their symbol tables than their user values.